Core/clim-core/redefinition.lisp

```diff
@@ -0,0 +1,176 @@
+;;;; (C) Copyright 2019 Jan Moringen
+;;;;
+;;;; This library is free software; you can redistribute it and/or
+;;;; modify it under the terms of the GNU Library General Public
+;;;; License as published by the Free Software Foundation; either
+;;;; version 2 of the License, or (at your option) any later version.
+;;;;
+;;;; This library is distributed in the hope that it will be useful,
+;;;; but WITHOUT ANY WARRANTY; without even the implied warranty of
+;;;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+;;;; Library General Public License for more details.
+;;;;
+;;;; You should have received a copy of the GNU Library General Public
+;;;; License along with this library; if not, write to the
+;;;; Free Software Foundation, Inc., 59 Temple Place - Suite 330,
+;;;; Boston, MA  02111-1307  USA.
+
+(in-package #:clim-internals)
+
+;;; Frame class redefinition notification protocol
+;;;
+;;; This is used for notifying application frame classes after aspects
+;;; such as the class itself, the `generate-panes' method or the
+;;; command table have been redefined. The notification sequence
+;;; begins with a call to `note-frame-class-redefined' which either
+;;; directly calls `handle-frame-class-redefinition' or schedules the
+;;; call to be made when leaving a
+;;; `with-delayed-redefinition-notifications'.
+;;;
+;;; This protocol is probably most commonly initiated by re-evaluating
+;;; the `define-application-frame' form that defined the frame class
+;;; in question, but any event that causes the `reinitialize-instance'
+;;; method of the class to execute also triggers it. In the context of
+;;; `define-application-frame',
+;;; `with-delayed-redefinition-notifications' delays the
+;;; `note-frame-class-redefinition' call until after all aspects
+;;; (`generate-panes' method, command table definition, etc) have been
+;;; executed.
+
+(defgeneric handle-frame-class-redefinition (class)
+  (:method ((class t))))
+
+(defvar *delayed-redefinition-notifications*)
+
+(defun call-with-delayed-redefinition-notifications (thunk)
+  (if (boundp '*delayed-redefinition-notifications*)
+      (funcall thunk)
+      (let ((*delayed-redefinition-notifications* '()))
+        (funcall thunk)
+        (map nil #'handle-frame-class-redefinition
+             *delayed-redefinition-notifications*))))
```

Comment:
not that it matters much, but shouldn't it be `(multiple-value-prog1 (funcall thunk) (mapc …))` (or in the first arm of if `(prog1 nil (funcall thunk))`)

Reply:
Yes, I think your suggestion makes it better. Changed. Although the only user, `define-application-frame` doesn't seem to have a specified return value.